test/test_puma_server_partial_hijack.rb

```diff
@@ -0,0 +1,136 @@
+require_relative "helper"
+require "puma/events"
+require "puma/server"
+require "net/http"
+require "nio"
+require "ipaddr"
+
+class TestPumaServerPartialHijack < Minitest::Test
+  parallelize_me!
+
+  def setup
+    @host = "127.0.0.1"
+
+    @ios = []
+
+    @app = ->(env) { [200, {}, [env['rack.url_scheme']]] }
+
+    @log_writer = Puma::LogWriter.strings
+    @events = Puma::Events.new
+  end
+
+  def teardown
+    @server.stop(true)
+    assert_empty @log_writer.stdout.string
+    assert_empty @log_writer.stderr.string
+
+    # Errno::EBADF raised on macOS
+    @ios.each do |io|
+      begin
+        io.close if io.respond_to?(:close) && !io.closed?
+        File.unlink io.path if io.is_a? File
+      rescue Errno::EBADF
+      ensure
+        io = nil
+      end
+    end
+  end
+
+  def server_run(**options, &block)
+    options[:log_writer]  ||= @log_writer
+    options[:min_threads] ||= 1
+    @server = Puma::Server.new block || @app, @events, options
+    @port = (@server.add_tcp_listener @host, 0).addr[1]
+    @server.run
+    sleep 0.15 if Puma.jruby?
+  end
+
+  # only for shorter bodies!
+  def send_http_and_sysread(req)
+    send_http(req).sysread 2_048
+  end
+
+  def send_http_and_read(req)
+    send_http(req).read
+  end
+
+  def send_http(req)
+    new_connection << req
+  end
+
+  def new_connection
+    TCPSocket.new(@host, @port).tap {|sock| @ios << sock}
+  end
+
+  def test_101_body
+    headers = {
+      'Upgrade' => 'websocket',
+      'Connection' => 'Upgrade',
+      'Sec-WebSocket-Accept' => 's3pPLMBiTxaQ9kYGzzhZRbK+xOo=',
+      'Sec-WebSocket-Protocol' => 'chat'
+    }
+
+    body = -> (io) {
+      io.wait_readable 0.1 # TRUFFLE
+      io.syswrite io.sysread(256)
+      io.close
+    }
+
+    server_run do |env|
+      [101, headers, body]
+    end
+
+    sock = new_connection
+    sock.syswrite "GET / HTTP/1.1\r\n\r\n"
+    resp = sock.sysread 1_024
+    echo_msg = "This should echo..."
+    sock.syswrite echo_msg
+
+    assert_includes resp, 'Connection: Upgrade'
+    assert_equal echo_msg, sock.sysread(256)
+  end
+
+  def test_101_header
+    headers = {
+      'Upgrade' => 'websocket',
+      'Connection' => 'Upgrade',
+      'Sec-WebSocket-Accept' => 's3pPLMBiTxaQ9kYGzzhZRbK+xOo=',
+      'Sec-WebSocket-Protocol' => 'chat',
+      'rack.hijack' => -> (io) {
+        io.wait_readable 0.1 # TRUFFLE
```

Comment:
Could you expand this comment just a little bit more? Why does Truffle need this?

Reply:
Updated comment.

Threads with JRuby/Truffle are different than MRI Rubies.  Thread creation/switching may take a bit more time when using JRuby/Truffle, especially with parallel testing in a multi VM CI environment.

Sometimes code additions like these may not be necessary when testing locally, other additions are required to pass both locally and in cloud CI.  A few tests use `parallelize_me! if ::Puma.mri?` because of these timing issues.

Re this particular issue, the following line
(`io.syswrite io.sysread(256)`) threw
`Read: #<Errno::EAGAIN: Resource temporarily unavailable>`